solver.go

```diff
@@ -361,34 +361,43 @@ func (s *solver) solve() (map[atom]map[string]struct{}, error) {
 		// that has yet to be selected.)
 		if awp, is := s.sel.selected(bmi.id); !is {
 			s.mtr.push("new-atom")
-			// Analysis path for when we haven't selected the project yet - need
-			// to create a version queue.
-			queue, err := s.createVersionQueue(bmi)
-			if err != nil {
-				// Err means a failure somewhere down the line; try backtracking.
-				s.traceStartBacktrack(bmi, err, false)
+
+			if !bmi.fromRoot {
+				// Here we have a flat package, and we don't need to check any dependencies
+				// for it. There is nothing to check here, so remove it.
+				s.unsel.remove(bmi)
```

Comment:
Oh wait, no, it's right. It's checking `rpm[pkg] == 1` because the map that's returned tells us the _number_ of things importing on a particular pkg in the current graph.

So, anyway. Because we have that check there, it means that we only add a bmi to the unselected queue if it would add to the package set required from a given project. If there's only one package in a project, and the project's already been pulled in, then that single package has necessarily already been selected, meaning that no new bmi can enter the unselected queue...so the case you're guarding against up in the main solve loop can't occur.

I hope it was vaguely possible to follow that :)